crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+/// the original owner.
+/// 
+/// Once the number of votes is reached the proposer of the upgrade can perform
+/// the upgrade.
+/// 
+/// Events are emitted to track the main operations on the proposal.
+/// A proposed upgrade lifetime is tracked via the 4 events:
+/// `UpgradeProposed`, `UpgradeVoted` and `UpgradePerformed` or `UpgradeDiscarded`.
+/// Also `Ballot`s are tracked via `BallotTransfered` and `BallotDeleted` in order to
+/// audit the lifetime of a vote.
+/// 
+/// Multiple upgrades can be live at the same time. That is not the expected behavior
+/// but there are no restriction to the number of upgrades open at any point in time.
+/// When that happens the first upgrade executed "wins" and subsequent attempt to
+/// authorize an upgrade will fail as the version will not match any longer.
+module sui::k_of_n_upgrade_policy {
+    use std::vector;
+    use sui::event;
+    use sui::object::{Self, ID, UID};
+    use sui::package::{Self, UpgradeCap, UpgradeTicket, UpgradeReceipt};
+    use sui::transfer;
+    use sui::tx_context::{Self, TxContext};
+    use sui::vec_set::{Self, VecSet};
+
+    /// The capability controlling the upgrade. 
+    /// Initialized with `new` is returned to the caller to be stored as desired.
+    struct KofNUpgradeCap has key, store {
+	    id: UID,
+        /// Upgrade cap of the package controlled by this policy.
+	    cap: UpgradeCap,
+        /// Number of votes required for the upgrade to be allowed.
+	    required_votes: u64,
+        /// Allowed voters. They will receive a ballot. They can be addresses or objects.
+        /// The creator of the policy will be added to the voters if not there already.
+	    voters: VecSet<address>,
+    }
+
+    /// A proposed upgrade that is going through voting.
+    /// `ProposedUpgrade` instances are shared objects that will be passed as 
+    /// an argument, together with a `Ballot`, when voting.
+    /// It's possible to have multiple proposed upgrades at the same time and 
+    /// the first successful update will obsolete all the others, given
+    /// an attempt to upgrade with a "concurrent" one will fail because of
+    /// versioning.
+    struct ProposedUpgrade has key {
+        id: UID,
+        /// The ID of the `KofNUpgradeCap` that this vote was initiated from.
+        cap: ID,
+        /// The address requesting permission to perform the upgrade.
+        /// This is the sender of the transaction that proposes and 
+        /// performs the upgrade.
+        signer: address,
+        /// The digest of the bytecode that the package will be upgraded to.
+        digest: vector<u8>,
+        /// The ballots allowed to vote for this upgrade.
+        allowed_voters: VecSet<ID>,
+        /// The current voters that have accepted the upgrade.
+        current_voters: VecSet<ID>,
+    }
+
+    /// A request to vote for an upgrade. Each possible voter will receive a `Ballot`
+    /// that they use to vote for the upgrade. 
+    struct Ballot has key {
```

Comment:
Reading over the rest of the protocol, I find the `Ballot` system a little difficult to follow.  My understanding is that the purpose of it is to allow a voter to delegate the right to vote, but only for a single vote.

- How important is this functionality? Could we get away with getting rid of it altogether?
- If not, how important is the ability to only delegate a single vote?  Would it be sufficient to allow the rights to vote to be transferred between addresses? This would involve transferring `VoterCap`-s to all the allowed voters when the upgrade cap is created, instead of a ballot when the vote is proposed.  These caps can be transferred between addresses, and whoever has one can vote on the proposal.

At the moment, the protocol is a bit of a strange mix of address- and cap-based authentication, so it would be good to lean fully one way or the other.

Reply:
I think it's a good point and the idea of transferring the vote may not be needed at all in this first iteration.
The principle was to create a link between vote and proposal in a way that I could ask other entities to vote in case they may be more competent for that specific proposal, sort of a consortium of reviewers. That is definitely not needed for starters, so I may review that assumption.
The `VoterCap` is interesting and somewhat orthogonal to the Ballot. I am curious to talk to @jnaulty about it to understand what we track from a "governance" perspective. In any case the `VoterCap` could solve a problem with addresses/entities becoming obsolete for that policy. So there is value in that.

Reply:
I removed `Ballot` in favor of `VotingCap` we can reintroduce `Ballot` at a later point.
`VotingCap` gives us a way to addresses that will be obsolete and we can add the functionality later